### PR TITLE
mgr/dashboard: add latency chart 

### PR DESCRIPTION
**Pull request by VallariAg**
Visualize "ceph osd perf" command with a latency chart on the dashboard's Performance section.  

Why did I choose this chart? To compare values of grouped items, line chart and bar chart (or similar charts like histogram) are two great options. To decide between the two, line chart showed the difference between the commit_latency and apply_latency more clearly than overlapping bar charts did. So, I decided to go for the line chart! 

Line chart:
image

Alternatively, Bar Chart (showing the exact same data as above):
image

Tests: 
One of dashboard's e2e frontend visual tests `visualTests/dashboard.vrt-spec.ts` is failing (because I made UI changes), please let me know how to update the tests for this!   

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by epuertat** [comment]
Applitools is failing but this time is a true change (new widget at home!):

se_c43ca5ff-28b2-41dd-8497-b4c6dc2ce183

**Comment by VallariAg** [comment]
@epuertat suggested to use bar charts over line charts with the following ideas:
- Sorting: instead of arranging by OSD ID, which is a rather random criterion (OSD 3 and 4 can be in different hosts, and OSD 2 could not exist), sort by the magnitude. You may sort by one of the individual latencies (apply) or the total sum.
- Top-k charts: basically displaying only the top-k (worst or best) columns
- To better describe all the latencies (and not only the worst cases), you can also display, for example, the 5-best columns and the average/median

For this, the widget would look something like this (where the first five are with worst latencies, a median, then best five latencies. All sorted by magnitude of total sum):
image

But now, in this approach, how would the chart look for less than 10 OSDs? 
Keeping two things in mind - both sides (worst and best) should have equal amount of bars and no repetition of data (an OSD cannot come up in both worst 5 and best 5 or be the median when it's in one of the two sections) - we can consider to do: 

- Option 1:  Vary the amount of bars on either side of the median. Example for 3-4 OSDs we can show “1 Worst - Median - 1 Best” (the image below) and for 5-6 OSDs "2 Worst - median - 2 Best" and so on. 
image

- Option 2: If there are 10 or fewer OSDs, we can show all OSDs latencies in a sorted manner. And for more than 10 OSDs, we can show the “Worst 5 - median - best 5” graph (where both sides are sorted as well). 
image

While option 1 is more consistent in design, the use of a median bar is little less meaningful when there are few OSDs. For example, in the 3 OSDs diagram above, we know that the median bar is the third remaining OSD, naming it by its OSD ID would be cleaner than "median <osd id>". So, it could be more helpful if we show all the latency values individually to a threshold count, and then show "best 5 - median - worst-5".

**Comment by pereman2** [comment]
jenkins retest this please

**Comment by pereman2** [comment]
Hey @VallariAg, what is missing in this PR?

**Comment by VallariAg** [comment]
> Hey @VallariAg, what is missing in this PR? 

Hey @pereman2,
I'm redoing the chart from Line Chart to one of the chart from the options that @epuertat mentioned in his last comment. 
I couldn't get it done due to my college exams, I'm planning to get this done by end of this week :)

File: src/pybind/mgr/dashboard/controllers/health.py [hunk 1]
```diff
@@ -92,6 +92,13 @@
             'up': (int, ''),
         }], '')
     }, ''),
+    'perf_stat': ({
+        'osd_perf': ([{
+            'osd': (int, ''),
+            'commit_latency_ms': (int, ''),
```
Comment (sunilangadi2) [inline src/pybind/mgr/dashboard/controllers/health.py:98]: Thanks ! @VallariAg This is a Nice improvement for OSD performance,
Can you please explain more about what is exactly _commit_latency_ and _apply_latency_ in osd's?
and I prefer Bar chart it looks good and easy to understand.

Comment (VallariAg) [reply src/pybind/mgr/dashboard/controllers/health.py:98]: Hey @sunilangadi2, 
When we carry out an operation like `rados bench -p testbench 10 write --no-cleanup`, the command `ceph osd perf` gives the following stats:
```
osd  commit_latency(ms)  apply_latency(ms)
  2                 154                154
  0                 143                143
  1                 168                168
```
Here, the commit latency is the time taken to commit the ongoing operation and the apply latency is the time taken to apply the changes to disks. 

Indeed, the bar chart looks like a good solution! I'm done with it's implementation, I'm currently looking into the grafana charts for maybe a more engaging solution.

**Comment by VallariAg** [comment]
**The bar charts with percentiles:**
For less than 5 OSDs, with the median and the 2-sigma percentiles (5th and 95th):
image 
And for more than 5 OSDs, with 3-sigma percentiles (1th, 5th 50th, 95th and 99th):
image

**Comment by pereman2** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard/health/health.component.html [hunk 2]
```diff
@@ -226,6 +226,18 @@
                   *ngIf="healthData.scrub_status">
       {{ healthData.scrub_status }}
     </cd-info-card>
+
+    <cd-info-card cardTitle="Latency"
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard/health/health.component.html:230]:
```suggestion
    <cd-info-card cardTitle="OSDs Latency"
```

File: src/pybind/mgr/dashboard/controllers/health.py [hunk 3]
```diff
@@ -157,6 +165,18 @@ def all_health(self):
...
+    def perf_stat(self):
+        osd_perf = []
+        osd_stats = mgr.get('osd_stats')
+
+        for osd in osd_stats['osd_stats']:
+            osd_perf += [{
+                'osd': osd['osd'],
+                'apply_latency_ms': osd['perf_stat']['apply_latency_ns'] / 1000000.0,
+                'commit_latency_ms': osd['perf_stat']['commit_latency_ns'] / 1000000.0,
+            }]
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/controllers/health.py:177]:
```suggestion
            osd_perf.append({
                'osd': osd['osd'],
                'apply_latency_ms': osd['perf_stat']['apply_latency_ns'] / 1000000.0,
                'commit_latency_ms': osd['perf_stat']['commit_latency_ns'] / 1000000.0,
            })
```

**Comment by pereman2** [comment]
There are linting errors:
```
[lint:tslint  ] ERROR: 109:18  no-trailing-whitespace  trailing whitespace
[lint:tslint  ] 
[lint:tslint  ] /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard/health/health.component.ts:73:18
[lint:tslint  ] ERROR: 73:18   no-trailing-whitespace  trailing whitespace
[lint:tslint  ] ERROR: 299:9   prefer-const            Identifier 'latency_count' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 302:11  prefer-const            Identifier 'fifth_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 303:11  prefer-const            Identifier 'median_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 304:11  prefer-const            Identifier 'ninty_fifth_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 306:11  no-shadowed-variable    Shadowed name: 'data'
[lint:tslint  ] ERROR: 306:11  prefer-const            Identifier 'data' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 312:23  quotemark               " should be '
[lint:tslint  ] ERROR: 312:29  quotemark               " should be '
[lint:tslint  ] ERROR: 312:36  quotemark               " should be '
[lint:tslint  ] ERROR: 313:27  no-trailing-whitespace  trailing whitespace
[lint:tslint  ] ERROR: 314:16  quotemark               " should be '
[lint:tslint  ] ERROR: 320:11  prefer-const            Identifier 'zero_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 320:28  no-inferrable-types     Type number trivially inferred from a number literal, remove type annotation
[lint:tslint  ] ERROR: 321:11  prefer-const            Identifier 'hundred_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 322:11  prefer-const            Identifier 'median_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 323:11  prefer-const            Identifier 'fifth_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 324:11  prefer-const            Identifier 'ninty_fifth_percentile' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 326:11  no-shadowed-variable    Shadowed name: 'data'
[lint:tslint  ] ERROR: 326:11  prefer-const            Identifier 'data' is never reassigned; use 'const' instead of 'let'.
[lint:tslint  ] ERROR: 334:23  quotemark               " should be '
[lint:tslint  ] ERROR: 334:29  quotemark               " should be '
[lint:tslint  ] ERROR: 334:35  quotemark               " should be '
[lint:tslint  ] ERROR: 334:42  quotemark               " should be '
[lint:tslint  ] ERROR: 334:49  quotemark               " should be '
[lint:tslint  ] ERROR: 335:27  no-trailing-whitespace  trailing whitespace
[lint:tslint  ] ERROR: 336:16  quotemark               " should be '
```

**Pull request closed without merging**